### PR TITLE
Feat(Compiler-Base): Add code snippet component

### PR DESCRIPTION
**Pull request by zong-zhe**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

issue #115.

#### 2. What is the scope of this PR (e.g. component or file name):

compiler_base/error/diagnostic/components.rs

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

add UnderLine, IndentWithPrefix, StringWithStyle, CodeSnippet in compiler_base/error/src/diagnostic/components.rs to show code context in diagnostic.

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 1]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+    /// # use compiler_base_error::DiagnosticStyle;
+    /// # use rustc_errors::styled_buffer::StyledBuffer;
+    ///
+    /// let mut sb = StyledBuffer::<DiagnosticStyle>::new();
+    /// let mut errs = vec![];
+    /// // If you want to add 3 indents and rendering text: "   ^"
+    /// let indent = IndentWithPrefix::new("^".to_string(), 3, None);
+    /// indent.format(&mut sb, &mut errs);
+    /// ```
+    pub fn new(prefix: String, indent: usize, prefix_style: Option<DiagnosticStyle>) -> Self {
```
Comment (Peefy) [inline kclvm/compiler_base/error/src/diagnostic/components.rs:179]: `prefix` or `prefix_label` or `label` ? Compare to line 263

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 2]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+    ///     lo: new_byte_pos(22),
+    ///     hi: new_byte_pos(42),
+    /// }.span();
+    ///
+    /// // 3. You can create the `CodeSnippet` by the `SourceFile`,
+    /// // and render text "Line 2 Code Snippet.".
+    /// let code_snippet = CodeSnippet::new_with_source_map(code_span, Arc::new(sm));
+    /// code_snippet.format(&mut sb, &mut errs);
+    /// ```
+    pub fn new_with_source_map(code_span: Span, source_map: Arc<SourceMap>) -> Self {
```
Comment (Peefy) [inline kclvm/compiler_base/error/src/diagnostic/components.rs:343]: s/new_with_source_map/new?

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 3]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
         }
     }
 }
+
+/// `StringWithStyle` is a component of diagnostic to display a string with style.
```
Comment (Peefy) [inline kclvm/compiler_base/error/src/diagnostic/components.rs:61]: s/StringWithStyle/StyledString? 

Besides, using `StyledString<T>` in rustc_errors?

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 4]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+impl StringWithStyle {
+    /// You can new a `StringWithStyle` with the string content and `DiagnosticStyle`.
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use compiler_base_error::components::StringWithStyle;
+    /// # use compiler_base_error::DiagnosticStyle;
+    /// let string_styled = StringWithStyle::new_with_style("A styled string".to_string(), Some(DiagnosticStyle::NeedFix));
+    /// ```
+    pub fn new_with_style(content: String, style: Option<DiagnosticStyle>) -> Self {
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 5]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
 }
+
+/// `StringWithStyle` is a component of diagnostic to display a string with style.
+pub struct StringWithStyle {
+    content: String,
+    style: Option<DiagnosticStyle>,
+}
+
+impl StringWithStyle {
+    /// You can new a `StringWithStyle` with the string content and `DiagnosticStyle`.
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 6]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+/// ```
+pub struct IndentWithPrefix {
+    indent: usize,
+    prefix: StringWithStyle,
+}
+
+const DEFAULT_INDENT_PREFIX_LABEL: &str = "|";
+
+impl IndentWithPrefix {
+    /// You can new a `IndentWithPrefix` by default label with 0 indent.
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 7]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+    /// ```rust
+    /// # use compiler_base_error::components::StringWithStyle;
+    /// # use compiler_base_error::DiagnosticStyle;
+    /// let string_styled = StringWithStyle::new_with_style("A styled string".to_string(), Some(DiagnosticStyle::NeedFix));
+    /// ```
+    pub fn new_with_style(content: String, style: Option<DiagnosticStyle>) -> Self {
+        Self { content, style }
+    }
+
+    /// You can new a `StringWithStyle` with no style.
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 8]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+    /// indent.format(&mut sb, &mut errs);
+    /// ```
+    pub fn default() -> Self {
+        Self {
+            indent: 0,
+            prefix: StringWithStyle::new_with_no_style(DEFAULT_INDENT_PREFIX_LABEL.to_string()),
+        }
+    }
+
+    /// You can new a `IndentWithPrefix` by default label with custom indents.
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs [hunk 9]
```diff
@@ -54,3 +57,350 @@ impl Component<DiagnosticStyle> for Label {
...
+
+    /// You can new a `StringWithStyle` with no style.
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use compiler_base_error::components::StringWithStyle;
+    /// # use compiler_base_error::DiagnosticStyle;
+    /// let string_styled = StringWithStyle::new_with_no_style("A styled string".to_string());
+    /// ```
+    pub fn new_with_no_style(content: String) -> Self {
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs (hunk 2 above)
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs (hunk 3 above)
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs]: fixed

File: kclvm/compiler_base/error/src/diagnostic/components.rs (hunk 1 above)
Comment (zong-zhe) [reply kclvm/compiler_base/error/src/diagnostic/components.rs:156]: Fixed

**Review by ldxdl — DISMISSED** [review]
LGTM

**Review by Peefy — APPROVED** [review]
LGTM

**Review by He1pa — APPROVED** [review]
LGTM

**Pull request merged**